apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/EpochTickClock.java

```diff
@@ -0,0 +1,48 @@
+package co.elastic.apm.impl.transaction;
+
+import co.elastic.apm.objectpool.Recyclable;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This clock makes sure that each {@link Span} and {@link Transaction} uses a consistent clock
+ * which does not drift in case of NTP updates or leap seconds.
+ * <p>
+ * The clock is initialized with wall clock time when the transaction starts and from there on
+ * uses {@link System#nanoTime()} in order to be able to measure durations.
+ * {@link System#currentTimeMillis()}, which uses wall clock time is not suitable for measuring durations because of the aforementioned
+ * possibility of clock drifts.
+ * </p>
+ */
+public class EpochTickClock implements Recyclable {
+
+    private long nanoTimeOffsetToEpoch;
+
+    public void init(EpochTickClock other) {
+        this.nanoTimeOffsetToEpoch = other.nanoTimeOffsetToEpoch;
+    }
+
+    /**
+     * Initializes and calibrates the clock based on wall clock time
+     */
+    public void init() {
+        init(System.currentTimeMillis(), System.nanoTime());
+    }
+
+    public void init(long epochMillis, long nanoTime) {
+        nanoTimeOffsetToEpoch = nanoTime - TimeUnit.MILLISECONDS.toNanos(epochMillis);
```

Comment:
That's correct, but got me confused, as it is not really an offset (it yields a negative value). No need to change for correctness, but I would say it is more readable if you reverse the subtraction order to get an offset to epoch, and then add it to the nano time in [getEpochMicros](https://github.com/elastic/apm-agent-java/pull/232/files#diff-dcb41b5d39f04a99247e0d7aad3ab48bR41) instead of subtracting